api/agent/drivers/docker/imageCache.go

```diff
@@ -0,0 +1,108 @@
+package docker
+
+import (
+	"container/list"
+	"sync"
+
+	d "github.com/fsouza/go-dockerclient"
+)
+
+type ImageEvictor func(d.APIImages) error
+
+// Cache is an LRU cache, safe for concurrent access.
+type Cache struct {
+	totalSize int64
+	mu        sync.Mutex
+	ll        *list.List
+	cache     map[string]*list.Element
+	maxSize   int64
+	onEvict   ImageEvictor
+}
+
+// *entry is the type stored in each *list.Element.
+type entry struct {
+	key   string
+	value d.APIImages
+}
+
+// New returns a new cache with the provided maximum items.
+func NewLRU(maxSize int64, onEvict ImageEvictor) *Cache {
+	return &Cache{
+		ll:      list.New(),
+		cache:   make(map[string]*list.Element),
+		onEvict: onEvict,
+	}
+}
+
+// Add adds the provided key and value to the cache, evicting
+// an old item if necessary.
+func (c *Cache) Add(key string, value d.APIImages) {
+	c.mu.Lock()
+
+	// Already in cache?
+	if ee, ok := c.cache[key]; ok {
+		c.ll.MoveToFront(ee)
+		c.totalSize = (c.totalSize - ee.Value.(d.APIImages).Size) + value.Size
+		ee.Value.(*entry).value = value
+		return
+	}
+
+	// Add to cache if not present
+	ele := c.ll.PushFront(&entry{key, value})
+	c.cache[key] = ele
+	c.totalSize += value.Size
+	c.mu.Unlock()
+
+	for c.TotalSize() > c.maxSize {
```

Comment:
should probably be a loop here as we don't know single evict will satisfy the total size <= max size requirement.